### PR TITLE
Concatenate directly into shared memory when constructing batches

### PR DESCRIPTION
**Pull request by colesbury**
This saves an extra memory copy, which speeds up data loading a bit
(5-10% with accimage).

As part of this change:

 * torch.cat accepts keyword argument out
 * sepcifiying out=None is treated like not specifying out

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -75,7 +82,14 @@ def _pin_memory_loop(in_queue, out_queue, done_event):
 def default_collate(batch):
     "Puts each data field into a tensor with outer dimension batch size"
     if torch.is_tensor(batch[0]):
-        return torch.stack(batch, 0)
+        out = None
+        if _use_shared_memory:
+            # If we're in a background process, concatenate directly into a
+            # shared memory tensor to avoid an extra copy
+            numel = sum([x.numel() for x in batch])
```
Comment (apaszke) [inline torch/utils/data/dataloader.py]: you can remove the brackets and give an iterator to `sum`

**Review by apaszke — APPROVED** [review]
LGTM 👍

**Comment by chenyuntc** [comment]
seems to cause error
```
import torch as t
a=t.Tensor(2,4)
b=t.Tensor(2,4)
c=t.Tensor(2,8)
t.cat(a,b,out=c)
```
```python
TypeError                                 Traceback (most recent call last)
<ipython-input-1-ce3d61609aec> in <module>()
      3 b=t.Tensor(2,4)
      4 c=t.Tensor(2,8)
----> 5 t.cat(a,b,out=c)

TypeError: cat received an invalid combination of arguments - got (torch.FloatTensor, torch.FloatTensor, out=torch.FloatTensor), but expected one of:
 * (sequence[torch.FloatTensor] seq)
 * (sequence[torch.FloatTensor] seq, int dim)
```

also see  https://discuss.pytorch.org/t/cat-got-an-unexpected-keyword-argument-out/2151

**Comment by colesbury** [comment]
cat expects a sequence and a dimension : torch.cat([a, b], dim, out=c)

**Comment by boeddeker** [comment]
@colesbury Is there a reason why numpy arrays in `default_collate` do not use shared memory?
I would suggest changing
https://github.com/pytorch/pytorch/blob/e8754ee017d8f6eb047456f7b202e42ba3501fa6/torch/utils/data/dataloader.py#L218
to
```python
return default_collate([torch.from_numpy(b) for b in batch])
```

**Comment by colesbury** [comment]
@boeddeker that seems fine. Can you send a PR with the change?

**Comment by boeddeker** [comment]
ok, I opened a PR

**Pull request merged**